client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java

```diff
@@ -0,0 +1,153 @@
+ * limitations under the License.
+ */
+package org.kaaproject.kaa.client.configuration.base;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.kaaproject.kaa.client.KaaClientProperties;
+import org.kaaproject.kaa.client.configuration.ConfigurationHashContainer;
+import org.kaaproject.kaa.client.configuration.ConfigurationProcessor;
+import org.kaaproject.kaa.client.configuration.storage.ConfigurationStorage;
+import org.kaaproject.kaa.common.hash.EndpointObjectHash;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public abstract class AbstractConfigurationManager implements ConfigurationManager {
+    private static final Logger LOG = LoggerFactory.getLogger(AbstractConfigurationManager.class);
+
+    private final Set<ConfigurationListener> listeners = new HashSet<ConfigurationListener>();
+    private final KaaClientProperties properties;
+    protected final ConfigurationDeserializer deserializer = new ConfigurationDeserializer();
+
+    private byte[] configurationData;
+    private ConfigurationStorage storage;
+    private ConfigurationHashContainer container;
+
+    public AbstractConfigurationManager(KaaClientProperties properties) {
+        super();
+        this.properties = properties;
+        container = new HashContainer();
+    }
+    
+    @Override
+    public void init() {
+        getConfigurationData();
+    }
+
+    @Override
+    public boolean addListener(ConfigurationListener listener) {
+        if (listener != null) {
+            LOG.trace("Adding listener {}", listener);
+            synchronized (listeners) {
+                return listeners.add(listener);
+            }
+        } else {
+            throw new RuntimeException("Can't add null as a listener");
+        }
+    }
+
+    @Override
+    public boolean removeListener(ConfigurationListener listener) {
+        synchronized (listeners) {
+            return listeners.remove(listener);
+        }
+    }
+
+    @Override
+    public ConfigurationProcessor getConfigurationProcessor() {
+        return new ConfigurationProcessor() {
+
+            @Override
+            public void processConfigurationData(ByteBuffer buffer, boolean fullResync) throws IOException {
+                if (fullResync) {
+                    configurationData = toByteArray(buffer);
+                    if (LOG.isTraceEnabled()) {
+                        LOG.trace("Received configuration data {}", Arrays.toString(configurationData));
+                    }
+                    if (storage != null) {
+                        LOG.debug("Persisting configuration data from storage {}", storage);
+                        storage.saveConfiguration(ByteBuffer.wrap(configurationData));
+                        LOG.debug("Persisted configuration data from storage {}", storage);
+                    }
+                    for (ConfigurationListener listener : listeners) {
```

Comment:
I think it may cause a race condition between add/remove and processConfigurationData. Doesn't it?